[OOP - 로또] 2단계 미션 제출합니다.

bigpel66 opened the pull request:
1단계 미션의 마지막 피드백으로 주셨던 2가지 내용인

1) stream.count() 를 써도 되지 않을까요 ?
2) lotto에 contains를 구현해주면 굳이 Stream을 안돌아도 될 것 같아요 ?

에 대해서 반영하였습니다.

1에 대해선... count()라는 스트림 메서드를 이용해본 경험이 적었는데, 리뷰 덕분에 자잘하지만 더 적합한 메서드를 알게 되었던 거 같습니다.
2에 대해선... Lotto에 container를 두고 포함 여부 및 일치 개수를 판별하도록 하니, WinningLotto의 로직이 보다 깔끔하게 정리된 거 같습니다.

1단계 피드백에 감사드리며, 이번 미션 리뷰도 잘 부탁드립니다.

감사합니다.

dongjoon1251 commented on src/main/java/mission/lotto/controller/LottoController.java:
```diff
@@ -2,62 +2,50 @@
-    private TryNumber tryNumber;
-    private Lottos lottos;
+    public static void main(String[] args) {
+        LottoController lc = new LottoController();
```
LottoController 객체를 따로 만들지 않고 해당 클래스 내부에서 사용하는 메소드는 static으로 정의해서 사용해도 무방합니다.

dongjoon1251 commented on src/main/java/mission/lotto/controller/LottoController.java:
```diff
@@ -2,62 +2,50 @@
-    private Lottos lottos;
+    public static void main(String[] args) {
+        LottoController lc = new LottoController();
+        UserMoney userMoney;
```
선언과 할당이 분리되어 있는데 합쳐도 괜찮을 것 같습니다.

dongjoon1251 commented on src/main/java/mission/lotto/domain/Lottos.java:
```diff
@@ -11,23 +18,18 @@ public Lottos(List<Lotto> lottos) {
-        }
-        return result;
+    public static Lottos buyLottos(int manualCount, List<List<Integer>> manualInput, UserMoney userMoney) {
+        int totalQuantity = lottoAmount(userMoney);
```
static 생성자를 잘 만들어 주셨네요 👍 
totalQuantity - manualCount 이부분으로 인해 예외가 빌생할 수 있을 것 같습니다. 검증 로직을 추가해보는게 어떨까요?

dongjoon1251 replied:
해당 메소드에 대한 테스트 케이스도 추가해보는게 어떨까요?

kth496 replied:
테스트해보니 실제로 해당 로직에서 문제가 발생하네요 😭 코드를 수정하고 테스트 코드로 추가하겠습니다!

dongjoon1251 commented on src/main/java/mission/lotto/domain/Lottos.java:
```diff
@@ -11,23 +18,18 @@ public Lottos(List<Lotto> lottos) {
         Collections.sort(this.lottos);
```
생성자는 private 으로 바꿔서 내부에서만 접근하게 하고 생성은 static method로 모두 하도록 바꿔보면 어떨까요?
생성자에서는 할당만 하고 sort는 이전에 처리해서 넣어주면 어떨까요?

dongjoon1251 commented on src/main/java/mission/lotto/domain/Number.java:
```diff
@@ -1,21 +1,34 @@
+
+    public static Number of(int number) {
         if (isInvalidNumber(number)) {
             throw new IllegalArgumentException("숫자는 1~45 에 포함 되어야 합니다");
```
예외 클래스를 직접 만들어서 적용해보면 어떨까요?

kth496 replied:
1단계에서 말씀해주신 부분인데 빼먹었네요 😭😭

dongjoon1251 replied:
"숫자는 1~45 에 포함 되어야 합니다" 메시지에 들어있는 숫자도 상수를 사용하면 어떨까요?

dongjoon1251 commented on src/main/java/mission/lotto/domain/Lotto.java:
```diff
@@ -4,47 +4,70 @@
+                .map(Number::of)
+                .collect(Collectors.toSet());
+
+        if (checkNumbersLength(lottoNums)) {
```
사용된 메소드는 사용한 메소드 바로 아래에 두어서 가독성을 높여보면 어떨까요?

dongjoon1251 commented on src/main/java/mission/lotto/view/OutputView.java:
```diff
@@ -6,20 +6,26 @@
     }
 
     public static void resultView(Map<Rank, Integer> allLottoRank) {
         System.out.println("당첨 통계");
```
문자열은 모두 상수로 분리해보면 어떨까요?

kth496 replied:
상수로 분리하는 과정에서 빼억은 것이 있었네요 😭 수정하겠습니다 :)

dongjoon1251 commented on src/main/java/mission/lotto/controller/LottoController.java:
```diff
@@ -2,62 +2,50 @@
 
-    public void start() {
-        registerUserMoney(InputView.enterUerMoney());
+        userMoney = lc.registerUserMoney(InputView.enterUerMoney());
```
메소드로 분리하지 않고 바로 new UserMoney를 호출하게 하면 어떨까요?

kth496 replied:
과도하게 메서드 래핑을 했었네요.. 😭

dongjoon1251 commented on src/main/java/mission/lotto/controller/LottoController.java:
```diff
@@ -2,62 +2,50 @@
-        OutputView.autoBoughtView(lottos);
-
-        LottoAnswer lottoAnswer = makeLottoAnswer(
+        LottoAnswer lottoAnswer = lc.makeLottoAnswer(
```
LottoAnswer를 생성하는 로직을  LottoAnswer 내부로 옮기고 여기도 new  LottoAnswer만으로 표현해보면 어떨까요?

dongjoon1251 commented on src/main/java/mission/lotto/util/Statistics.java:
```diff
@@ -0,0 +1,32 @@
+import java.util.TreeMap;
+
+public class Statistics {
+
```
두 메소드 모드 Lottos, LottoAnswer를 파라미터로 받고 있네요. 생성자로 처음에 한번 받아서 필요한 Map<Rank, Integer> result 를 가지는 일급 컬렉션으로 리팩토링해보면 어떨까요? :)

kth496 replied:
생각지도 못했었네요.. 코드 중복을 많이 줄일 수 있는 방향이라고 생각합니다 😃

dongjoon1251 commented on src/main/java/mission/lotto/domain/Number.java:
```diff
@@ -1,21 +1,34 @@
     private final int number;
-    private static final int MIN_NUMBER_VALUE = 1;
-    private static final int MAX_NUMBER_VALUE = 45;
+    public static final int MIN_NUMBER_VALUE = 1;
```
상수는 항상 맨 위에 정의해주세요 :)

dongjoon1251 commented on src/main/java/mission/lotto/controller/LottoController.java:
```diff
@@ -2,62 +2,50 @@
+        OutputView.totalEarningsView(lc.getRateOfProfit(lottos, lottoAnswer, userMoney));
     }
 
-    private void buyLottos(NumGenerator numberGenerator) {
```
NumGenerator 인터페이스는 사용되는 부분이 없는데 삭제하면 어떨까요?

dongjoon1251 commented on src/main/java/mission/lotto/domain/Rank.java:
```diff
@@ -19,12 +21,15 @@
+                .filter(rank -> rank.isSameCorrectNo(correctNo))
+                .filter(rank -> !rank.equals(SECOND) || hasBonus)
+                .findFirst()
+                .orElse(UNRANKED);
```
correctNo에 7이상의 값이 들어올 경우에도 UNRANKED로 간주하는게 좋을까요? 한번 고민해주시기 바랍니다.
그리고 변수명은 되도록 축약형은 지양해주시면 좋을 것 같습니다.

kth496 replied:
`UNRANKED`를 아무 의미 없다는 표현으로 작성했습니다. `Lotto` 클래스의 생성자에서 중복없는 6개 숫자를 받아오기 때문에 7이상의 수가 들어오지 않는다고 생각했네요.. 문제가 될 수 있는 부분이 있을까요??

dongjoon1251 replied:
현재 요구사항만 봤을 때는 문제가 없을 수 있고 구현해주신 방법으로 구현해도 무방합니다. 
다만, 해당 메소드가 public으로 구현되어 있어서 다른 동료가 작업할 때, 7 이상의 숫자를 입력해도 문제가 없는 것을 보고 잘못된 상황이 아니라고 인지할 가능성이 있습니다.

dongjoon1251 commented on src/test/java/mission/lotto/domain/RankTest.java:
```diff
@@ -1,46 +1,42 @@
-    public void rankCreateTest2() {
-        Rank rank = Rank.getRank(5, true);
-        assertThat(rank).isEqualTo(Rank.SECOND);
+    static Stream<Arguments> underThirdGenerator() {
```
ParameterizedTest 를 잘 적용해주셨네요 💯 
다만, 성공하는 테스트만 작성된 것으로 보입니다. 실패하는 케이스는 어떤 경우들이 있을까요?

kth496 replied:
실패하는 경우도 추가해보겠습니다!

dongjoon1251 commented on src/main/java/mission/lotto/domain/LottoAnswer.java:
```diff
@@ -4,10 +4,10 @@
+    public LottoAnswer(Lotto answerNumbers, Number bonusNumber) {
         if (answerNumbers.isContainLottoNumber(bonusNumber)) {
             throw new IllegalArgumentException("당첨 번호와 보너스 번호가 같을 수 없습니다.");
         }
```
getAnswerNumbers 를 통해서  set을 전달할 때 Collections.unmodifiableSet()로 한번 감싸서 리턴해주면 어떤 이점이 있을까요? :)

kth496 replied:
불변 컬렉션으로 변환하면서 얻을 수 있는 이점이 많다고 생각합니다. 도메인 객체 이외의 곳에서 변경을 시도했을 때 예외를 발생시키므로 이를 방지할 수 있을 것 같아요. :)

dongjoon1251 reviewed (changes requested):
먼저 질문에 대한 답변을 드리자면 아래와 같습니다.

1. stream은 디버깅하기 어렵습니다. 그래서 최대한 stream의 map이나 filter에 사용되는 메소드들을 분리해서 해당 메소드들을 테스트하는 방식을 사용합니다. 최종 연산으로 리턴된 객체를 테스트 하는 방식도 많이 사용됩니다. 원하는 값이 나오는지 중간에 map이나 forEach로 log를 찍거나 println하기도 합니다
2. 지금 작성해주신 대로 해주셔도 무방합니다 👍 
3. 잘 적용해주셨습니다 💯 

이전 피드백도 잘 반영해주셨네요 ㅎㅎ
추가로 몇가지 코멘트를 드렸습니다. 
모든 코드에 표시하지는 않았지만 getter를 통해서 list나 set을 리턴할 때 그냥 리턴하게 되면 어떤 위험성이 있을지 고민해보시면 좋을 것 같습니다.
그리고 성공하는 테스트들은 많이 작성되었지만 실패할 수 있는 케이스에 대해서도 테스트를 더 추가해보면 어떨까요?

kth496 commented:
안녕하세요! 코드 리뷰 해주신 내용을 바탕으로 리팩터링을 해 보았습니다. 말씀해주신 방향과 키워드를 바탕으로 고민하면서 몰랐던 자바 api도 알게되었습니다 😃 리뷰 중에서 `Rank`와 관련된 피드백은 제가 잘 이해하지 못해서 코멘트를 남겨두었으니 확인 부탁드립니다! 감사합니다 :)

dongjoon1251 commented on src/main/java/mission/lotto/domain/Lotto.java:
```diff
@@ -1,41 +1,48 @@
+                .collect(Collectors.toSet());
+
+        if (checkNumbersLength(lottoNums)) {
+            throw new IllegalArgumentException("로또는 중복 없는 6자리만 가능합니다.");
```
직접 만든 예외클래스 적용을 코드 전체적으로 적용하는게 어떨까요?
공통 예외클래스인 LottoException을 사용하기 보다는 에러 종류별로 만들고 메시지는 해당 예외클래스 내부에서 관리해보면 어떨까요?

dongjoon1251 commented on src/main/java/mission/lotto/domain/Lotto.java:
```diff
@@ -1,41 +1,48 @@
         Number bonus = answer.getBonusNumber();
 
-        boolean hasBonusNo = numbers.isContainLottoNumber(bonus);
+        boolean hasBonusNo = isContainLottoNumber(bonus);
```
hasBonusNo 와 같은 축약어도 지양해주세요:)

dongjoon1251 commented on src/main/java/mission/lotto/domain/Lotto.java:
```diff
@@ -1,41 +1,48 @@
-        combine.addAll(this.numbers.getNumbers());
+        combine.addAll(numbers);
 
         int correctNo = 12 - combine.size();
```
12는 어떤 의미가 있는 숫자인지 상수로 표현하면 어떨까요?

dongjoon1251 commented on src/main/java/mission/lotto/domain/Lottos.java:
```diff
@@ -1,37 +1,44 @@
         }
-        return result;
+
+        Stream<Lotto> manualLottos = IntStream.range(0, manualCount)
```
manaulLottos, autoLottos등을 생성하는 private 메소드로 분리해보면 어떨까요?

dongjoon1251 commented on src/main/java/mission/lotto/util/Statistics.java:
```diff
@@ -0,0 +1,39 @@
+
+    public Statistics(Lottos lottos, LottoAnswer lottoAnswer) {
+        Map<Rank, Integer> lottoRanks = new TreeMap<>();
+        for (Rank lotteryWinnings : Rank.values()) {
```
for문 대신 stream()을 사용해서 Map을 생성해보는 부분을 공부해보시면 좋을 것 같습니다 :)

dongjoon1251 reviewed (approved):
피드백들을 잘 반영해주셨네요 👍 
이번 미션은 여기서 마무리하도록 하겠습니다. 
몇가지 추가로 코멘트를 남겨드렸는데 확인해보시고 궁금하신 점은 따로 DM을 주셔도 됩니다. 
고생하셨습니다.

Commit:
feat: 수동 입력이 가능하도록 View를 구성
- ConsoleView의 입력을 받는 기능과 출력 메시지를 작성
- LottoGame이 수동 입력에 따른 복권 발급 가능 여부를 검증
- 자동과 수동을 나누어 입력 받은 리스트를 병합하여 게임을 기존 로직대로 진행

Commit:
refactor: Input 값이 잘못 입력된 경우의 복구를 가능하도록 수정
- ConsoleView와 Lotto의 입력 값 검증을 수행

Commit:
refactor: 2차 피드백 내용을 반영
- WinningLotto의 판별 로직 중 mapToInt -> sum을 count로 대체
- WinningLotto의 당첨 및 보너스 판별 로직의 책임 일부를 Lotto의 멤버를 이용하는 형태로 풀어냄

shared-moon reviewed (approved):
종환님 안녕하세요!
2단계 미션 잘 진행 해 주셨네요 :)
앞에서 구조를 잘 잡아주셔서 큰 변경 없이 미션 진행 할 수 있었던 것 같아요.
피드백을 좀 남겨 두었는데, 다음 미션 진행 전 확인 부탁드려요!
고생 많으셨습니다 :)
